working/static metaprogramming/intro.md

```diff
@@ -0,0 +1,421 @@
+    buildCode();
+  }
+}
+```
+
+The widget name is repeated six times, `fixedStuff` appears twice, and the
+remaining code is almost pure boilerplate that is identical across all
+`StatefulWidget` and `State` subclasses.
+
+There are other patterns in Flutter like this that are commonly applied to
+widget classes but require Flutter users to write a lot of repetitive code.
+Metaprogramming could enable the Flutter framework to express these patterns
+more tersely and with less chance for user error.
+
+## Requirements
+
+From these sample use cases and others, we can generalize some requirements:
+
+* **Must be able to introspect on the existing program**. Almost all the use
+  cases we want to cover need to be able to look at the fields of an existing
+  class at a minimum, as well as the fields of super classes in many cases. Many
+  also need to inspect the methods, constructors, and parameters of those.
+* **Must be able to synthesize new APIs in existing classes**. Most of the core
+  use cases require adding new methods to classes, whose signatures may not be
+  known ahead of time. Some also require the ability to add new fields, both
+  private and public to classes.
+* **Should be able to synthesize whole new classes**. Some of the core use cases
+  involve creating new classes from either other classes or even just top level
+  functions.
+* **Metaprogramming should be composable**. In one form or another,
+  metaprogramming applications should be able to be composed together. A data
+  class code generator for instance would ideally be a combination of several
+  more targeted code generators for each of the desired features (`copyWith()`,
+  `==`, `hashCode`, etc).
+* **Generated code must be debuggable and user visible**. See the Usability
+  section below.
+
+## Design constraints
+
+In addition to user-level requirements, there are some practical constraints to
+ensure that we don't compromise the existing advantages of Dart.
+
+* **Must support modular compilation**. Any individual library should be able to
+  be fully compiled (including running all macros) using only the transitive
+  dependencies of the library. To this end we should not expose any API or
+  functionality that allows global access to the program, outside of the current
+  library and/or its dependencies. Such functionality would defeat modular
+  compilation.
+* **Must support incremental compilation**. This is key to the hot reload
+  workflow—any change in the program should not have to rerun all
+  metaprogramming in the entire program. It should at worst require a rerun of
+  the metaprogramming in libraries that depend on the modified library.
+
+For the most part the design constraints here align with the modular compilation
+design constraints. In short, metaprogramming should not introduce global action
+at a distance.
+
+## Non-goals
+
+* **Modifying existing code**. We don’t see a need to allow modifiying existing
```

Comment:
Can you get more specific on what modifying existing code does and does not mean? Does it mean no deleting existing code but inserting arbitrary lines in-between existing code is fair game? If it ok to add conditionals that may cause blocks of existing code to not execute (for example turn off asserts)?

Reply:
Good point! I assumed the granularity of meta programming was at the member level. Add members, add classes, etc.

Would be good to clarify.

Reply:
By "Modify" we basically mean "Overwrite" - is that more clear?

@munificent are you fine with that wording change?

See the sentence here "Dart code that already has clearly defined meaning today should continue to mean the same thing after metaprogramming." that is the guiding principle. Allowing you to overwrite code clearly violates this principle (or at least would give you the power to violate it).

I don't think we want to restrict this purely to new top level and class declarations quite yet, but that may very well be where we end up (also likely allowing implementing abstract/external declarations).

Reply:
Thanks that is bit clearer. Does wrapping a a block of code with an if statement count as overwriting or not?

Reply:
Would this not support the use case of inlining then? Imagine I have some function, used 3 times, and it's better practice in flutter to have that inlined.

```
build(){
  return Column(
     _btn("1", onPressed: ...),
     _btn("2", onPressed: ...),
  )
}

@inline
Widget _btn(String label, onPressed) => Padding(padding: EdgeInsets.all(10), child: FlatButton(label, label, onPressed))
```
Could be turned into: 
```
build(){
  return Column(
    Padding(padding: EdgeInsets.all(10), child: FlatButton("1", label, onPressed)),
    Padding(padding: EdgeInsets.all(10), child: FlatButton("1", label, onPressed)),
  )
}
```
But that would require replacing the code inside the Widget tree.

It would basically get us Remi's functional widgets, without the annoying code generation phase:
https://pub.dev/packages/functional_widget

Reply:
> Would this not support the use case of inlining then? Imagine I have some function, used multiple times, and it's better performance in flutter to have that inlined.

Technically, that wouldn't violate the principles as long as the inlining maintained the same scoping rules for the inlined code as it had in its own function. This isn't really a use case we are targeting though I don't think, and the compiler already should be doing this where it makes sense.

> It would basically get us Remi's functional widgets, without the annoying code generation phase:

functional_widget is one of the use cases we have looked at and we are targeting something that could support that - but you don't need inlining for it. You really just need to be able to code generate a class IIRC.

Reply:
functional_widget is a weird one. It probably should not be done by code-generation, but function composition instead.

Code generation is far from ideal for such pattern. Because it's a very common mistake for users to use the function instead of the class, and it also causes naming issues.

Reply:
Ya it occured to me you could do this a different way, by making `_btn` be the class. But it seems like that could get quite complicated from a scope standpoint because the method will have access to all the scope of the containing stateful widget, including context, and any member vars, so the generated class would need to sniff out all those deps.  Where as if we just inline, everything would just kinda work.

Reply:
I think one thing that is being raised here is whether or not conditional compilation is covered by this proposal.  I think as stands, the answer might be "somewhat".  For example, I think this proposal is very compatible with a pattern where the user writes a static `MyClass.debugToString` method, and a macro generates a toString method which either calls that static method or not depending on compilation mode.  But I think we do not intend to support inclusion/exclusion of arbitrary fragments of text (a la #ifdef include) in the limit.

Reply:
There's a common conditional-compilation, use case we have where we have some native .io calls we want to make from our UI, and we want them to just be no-ops on Web (or anywhere else .io is not supported). 

It would be nice if this could support that. Currently we have to maintain 3 files, 1 interface, and 2 implementations. It would be nice if we could just maintain the 1file, and flag some methods, or an entire class, as no ops if .io is missing.

As a bit of an aside, it would be nice if this could ease conditional compilation in general. Right now it's quite a rough spot imo, very few junior to intermediate developers even understand how to set these up.

Unity's approach of
```
#if UNITY_STANDALONE
  // Do desktop stuff
#endif 
```
Is very handy, especially as C# allows you to inline the entire package path, keeping it all out of the compilers way if it's not supported. 
Like:
```
#if UNITY_EDITOR
  // Do stuff with UnityEditor.* namespace, which does not exist in a published build
#endif
```

This opens the door for the code we really want to write as developers: 
```
Future<double> getTitleBarHeight() async {
  #if CAN_USE_IO
      await io.methodChannel.invokeMethod("getTitlebarHeight") as double
  #else
     return 40;
  #endif
}
```
Which lets us just have a single class, share state/logic, and just swap out the offending lines on each platform. This is most maintainable, easiest to grok, and most productive, just from developer POV.
Cheers!